### PR TITLE
[SPARK-42557][CONNECT] Add Broadcast to functions

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, the connect functions missing the broadcast API. This PR want add this API to connect's functions.

### Why are the changes needed?
Add the broadcast function to connect's functions.scala.

### Does this PR introduce _any_ user-facing change?
 'No'.
New feature.

### How was this patch tested?
New test cases.

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -1228,6 +1228,22 @@ object functions {
...
+   * {{{
+   *   // left and right are DataFrames
+   *   left.join(broadcast(right), "joinKey")
+   * }}}
+   *
+   * @group normal_funcs
+   * @since 3.4.0
+   */
+  def broadcast[T](df: Dataset[T]): Dataset[T] = {
+    df.hint("broadcast")
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:1244]: is this enough? :)

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 2]
```diff
@@ -495,6 +495,14 @@ class ClientE2ETestSuite extends RemoteSparkSession {
     assert(joined.schema.catalogString === "struct<id:bigint,a:double>")
   }
 
+  test("broadcast join") {
+    val left = spark.range(100).select(col("id"), rand(10).as("a"))
+    val right = spark.range(100).select(col("id"), rand(12).as("a"))
+    val joined =
+      left.join(broadcast(right), left("id") === right("id")).select(left("id"), right("a"))
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:502]: can you check if we actually broadcast the right table?

**Review by hvanhovell — COMMENTED** [review]
Looks good! Can you add an check if actually broadcasting

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 3]
```diff
@@ -489,17 +489,31 @@ class ClientE2ETestSuite extends RemoteSparkSession {
...
   }
 
+  test("broadcast join") {
+    spark.conf.set("spark.sql.autoBroadcastJoinThreshold", "-1")
+    val left = spark.range(100).select(col("id"), rand(10).as("a"))
+    val right = spark.range(100).select(col("id"), rand(12).as("a"))
+    val joined =
+      left.join(broadcast(right), left("id") === right("id")).select(left("id"), right("a"))
+    assert(joined.schema.catalogString === "struct<id:bigint,a:double>")
+    testCapturedStdOut(joined.explain(), "BroadcastHashJoin")
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:513]: For later: we should have a better way to get the plan.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 4]
```diff
@@ -489,17 +489,31 @@ class ClientE2ETestSuite extends RemoteSparkSession {
...
+    spark.conf.set("spark.sql.autoBroadcastJoinThreshold", "10MB")
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:514]: Prefer to use try.. finally .. when you are resetting confs. That way failing tests do not start influencing others.

**Review by hvanhovell — APPROVED** [review]
LGTM

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 4 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:514]: Should we move `SQLHelper` to client module later? `withSQLConf` and other functions are more useful

Comment (beliefer) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:514]: @LuciferYang It's good idea. Let's do it later.

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by beliefer** [comment]
@LuciferYang I want support the similar `withSQLConf`.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 3 above)
Comment (beliefer) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:513]: Yeah.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:1244]: Yes.

**Comment by beliefer** [comment]
@LuciferYang Thank you.

**Comment by beliefer** [comment]
ping @HyukjinKwon @zhengruifeng @dongjoon-hyun

**Comment by hvanhovell** [comment]
Merging.

**Comment by beliefer** [comment]
@hvanhovell @LuciferYang Thank you.

**Pull request closed without merging**